### DISCUSSION TITLE
Load strategy conflict

### DISCUSSION THREAD

**Discussion by reuvenstr** [sqlalchemy/sqlalchemy#12021] [Usage Questions] [answered]
Hi

Im migrating to sqlalchemy 2.0 and found some issues with load strategies, for example:

```
import sqlalchemy as sa
from sqlalchemy.orm import relationship, joinedload, subqueryload, declarative_base, sessionmaker
from sqlalchemy.schema import ForeignKeyConstraint
from sqlalchemy import create_engine

Base = declarative_base()

class A(Base):
    __tablename__ = "a"

    id = sa.Column(sa.BigInteger(), primary_key=True, autoincrement=True)
    ats = relationship(
        "AA",
        secondary="e",
        secondaryjoin="e.c.id == aa.c.e_id",
        sync_backref=False,
        viewonly=True,
    )

class AA(Base):
    __tablename__ = "aa"

    id = sa.Column(sa.BigInteger(), primary_key=True, autoincrement=True)
    e_id = sa.Column(sa.BigInteger(), nullable=False)

    __table_args__ = (
        ForeignKeyConstraint(["e_id"], ["e.id"], ondelete="CASCADE"),
    )

class E(Base):
    __tablename__ = "e"

    id = sa.Column(sa.BigInteger(), primary_key=True, autoincrement=True)
    a_id = sa.Column(sa.BigInteger(), nullable=False)

    __table_args__ = (

            ForeignKeyConstraint(["a_id"], ["a.id"], ondelete="CASCADE"),
    )

DATABASE_URL = "sqlite:///example.db"
engine = create_engine(DATABASE_URL)

# Create all tables
Base.metadata.create_all(engine)

# Session setup
Session = sessionmaker(bind=engine)
session = Session()
session.query(A).options(joinedload(A.ats)).options(subqueryload(A.ats)).all()
```
And I'm facing this error `{InvalidRequestError}InvalidRequestError('Loader strategies for ORM Path[Mapper[A(a)] -> A.ats -> Mapper[AA(aa)]] conflict')`. It was ok in 1.4. 

You can ignore this error `{OperationalError}OperationalError('(sqlite3.OperationalError) no such column: e_1.a_id')`, it's only for example of multiple loading strategies on same attribute

Maybe there are some recommendations for such cases?

**Comment by zzzeek** [comment, marked as answer]
the error should be self-explanatory.   you have two different loader options for the same attribute and SQLAlchemy refuses to guess which one you want.    I have no idea which one in 1.4 would be chosen and you certainly wouldnt want to rely on that behavior in 1.4.

**Reply by reuvenstr** [reply]
Thanks for fast response!
So you mean it was broken in 1.4? As far as i understand its uses the last added option, but i can be mistaken. 
There is public method that can help me with cleaning previous options that were applied to query?
Its not straightforward like in example, there is some base query options, that are defined on super class, and child class is adding new options

**Reply by zzzeek** [reply]
it was undefined behavior, meaning, one or the other option would be favored but we had no documentation or tests to confirm what the order actually was. 

I think you should strive in your code to not have to "undo" loader options that you have applied to a query, just build the query once with the correct options.  try to revisit why you have a query with certain unwanted loader options applied to it in the first place and maybe revise that part of the code.

**Reply by reuvenstr** [reply]
Understood. Will try to fix such broken behavior 
Thanks a-lot!

**Reply by zzzeek** [reply]
im being a little hard-edged about it.   there could be an API or methodology to replace previous options of course.   what that does though is it means all of our APIs then seem like they should have "reversals", like what if I want to remove a WHERE clause or something like that, etc.     if options really are special in some way, that could provide justification.   but i am hoping that 2.0's improved strictness / clarity leads to people writing better code with less guessing up front
